tests/kokkos-based/helpers.hpp

```diff
@@ -84,5 +84,235 @@ auto create_stdvector_and_copy_rowwise(T sourceView)
+mdspan_t make_mdspan(const ValueType *data, std::size_t ext) {
+  return mdspan_t(data, ext);
+}
+
+template <typename ValueType>
+auto make_mdspan(std::vector<ValueType> &v) {
+  return make_mdspan(v.data(), v.size());
+}
+
+template <typename ValueType>
+auto make_mdspan(const std::vector<ValueType> &v) {
+  return make_mdspan(v.data(), v.size());
+}
+
+// create rank-2 mdspan (matrix)
+template <typename ValueType,
+          typename mdspan_t = typename _blas2_signed_fixture<ValueType>::mdspan_r2_t>
+mdspan_t make_mdspan(ValueType *data, std::size_t ext0, std::size_t ext1) {
+  return mdspan_t(data, ext0, ext1);
+}
+
+// no-tolerance (exact) comparison
+// TODO: add tolerance based comparison if needed (see is_same_matrix below)
+template <typename ElementType1,
+          typename LayoutPolicy1,
+          typename AccessorPolicy1,
+          typename ElementType2,
+          typename LayoutPolicy2,
+          typename AccessorPolicy2>
+bool is_same_vector(
+    const mdspan<ElementType1, extents<dynamic_extent>, LayoutPolicy1, AccessorPolicy1> &v1,
+    const mdspan<ElementType2, extents<dynamic_extent>, LayoutPolicy2, AccessorPolicy2> &v2)
+{
+  const auto size = v1.extent(0);
+  if (size != v2.extent(0))
+    return false;
+  const auto v1_view = KokkosKernelsSTD::Impl::mdspan_to_view(v1);
+  const auto v2_view = KokkosKernelsSTD::Impl::mdspan_to_view(v2);
+  int diff = false;
+  Kokkos::parallel_reduce(size,
+    KOKKOS_LAMBDA(const std::size_t i, decltype(diff) &d){
+        d = d || !(v1_view(i) == v2_view(i));
+	    }, diff);
+  return !diff;
+}
+
+template <typename ElementType1,
+          typename LayoutPolicy,
+          typename AccessorPolicy,
+          typename ElementType2>
+bool is_same_vector(
+    const mdspan<ElementType1, extents<dynamic_extent>, LayoutPolicy, AccessorPolicy> &v1,
+    const std::vector<ElementType2> &v2)
+{
+  return is_same_vector(v1, make_mdspan(v2));
+}
+
+template <typename ElementType1,
+          typename LayoutPolicy,
+          typename AccessorPolicy,
+          typename ElementType2>
+bool is_same_vector(
+    const std::vector<ElementType1> &v1,
+    const mdspan<ElementType2, extents<dynamic_extent>, LayoutPolicy, AccessorPolicy> &v2)
+{
+  return is_same_vector(v2, v1);
+}
+
+template <typename ElementType>
+bool is_same_vector(
+    const std::vector<ElementType> &v1,
+    const std::vector<ElementType> &v2)
+{
+  return is_same_vector(make_mdspan(v1), make_mdspan(v2));
+}
+
+// real diff: d = |v1 - v2|
+template <typename T, typename Enabled=void>
+class value_diff {
+public:
+  value_diff(const T &val1, const T &val2): _v(fabs(val1 - val2)) {}
+  operator T() const { return _v; }
+protected:
+  value_diff() = default;
+  T _v;
+};
+
+// real diff: d = max(|R(v1) - R(v2)|, |I(v1) - I(v2)|)
+// Note: returned value is of underlying real type
+template <typename T>
+class value_diff<std::complex<T>>: public value_diff<T> {
```

Comment:
[Optionally,] would you consider the following class hierarchy changes?

1. Make `_v` private.
2. Initialize `base` in the constructor, like this:

```c++
  value_diff(const std::complex<T> &val1, const std::complex<T> &val2) : base(base(val1.real(), val2.real()), base(val1.imag(), val2.imag())) {}
    const T dimag = base(val1.imag(), val2.imag());
    base::_v = std::max(dreal, dimag);
  }
```

I'm not a huge fan of `value_diff<complex<T>>` inheriting from `value_diff<T>`, but it seems to be working.

Reply:
Can I replace whole inheritance with much simpler and cleaner function:
```cpp
template <typename RealValue>
RealValue scalar_diff(RealValue v1, RealValue v2) {
  return std::abs(v2 - v1);
}

template <typename T>
T scalar_diff(const std::complex<T> &v1, const std::complex<T> &v2) {
  const auto dr = scalar_diff(v1.real(), v2.real());
  const auto di = scalar_diff(v1.imag(), v2.imag());
  return std::max(dr, di);
}

template <typename T>
T scalar_diff(const Kokkos::complex<T> &v1, const Kokkos::complex<T> &v2) {
  const auto dr = scalar_diff(v1.real(), v2.real());
  const auto di = scalar_diff(v1.imag(), v2.imag());
  return dr > di ? dr : di; // can't use std::max on GPU
}
```
?

Reply:
YES PLEASE : - )

Also, you might need to bless all but the `std::complex` overloads with `KOKKOS_INLINE_FUNCTION`, no?  Things might be different now, though.